### DISCUSSION TITLE
How to use ENTSOG client

### DISCUSSION THREAD

**Discussion by MakakWasTaken** [phax/phase4#83] [Q&A] [answered]
Hey, I have just recently started looking into AS4. So I am not that well-informed in the field.

For the past day or so, I have tried to implement a way to send a request with the AS4 protocol.
The problem I am running into is when I am trying to send a message, it gives me the following error:
`No certificates for user "null" were found for signature`

<details>
<summary>Full Stack:</summary>
<p>

  ```bash

  [main] INFO com.helger.phase4.mgr.MetaAS4Manager - MetaAS4Manager is initialized with in-memory data structures
  [main] INFO com.helger.phase4.mgr.MetaAS4Manager - Creating AS4 managers using factory class com.helger.phase4.mgr.ManagerFactoryInMemory
  [main] INFO com.helger.phase4.profile.AS4ProfileManager - 1 AS4 profile is registered 
  [main] INFO com.helger.phase4.mgr.MetaAS4Manager - MetaAS4Manager was initialized
  [main] INFO com.helger.phase4.sender.AS4BidirectionalClientHelper - Sending AS4 UserMessage to 'http://localhost:8080/as4' with max. 1 retries
  [main] INFO com.helger.phase4.wss.WSSConfigManager - None of the WSSConfig Security Providers is already installed - doing it now
  [main] INFO com.helger.phase4.messaging.crypto.AS4Signer - Now signing AS4 message. KeyIdentifierType=ISSUER_SERIAL; KeyAlias=null; SignAlgo=http://www.w3.org/2001/04/xmldsig-more#rsa-sha256; DigestAlgo=http://www.w3.org/2001/04/xmlenc#sha256; C14NAlgo=http://www.w3.org/2001/10/xml-exc-c14n#
  [main] ERROR com.helger.phase4.entsog.MainPhase4PeppolSenderLocalHost8080 - Error sending ENTSOG message via AS4
  com.helger.phase4.util.Phase4Exception: Wrapped Phase4Exception
	  at com.helger.phase4.entsog.Phase4ENTSOGSender$AbstractENTSOGUserMessageBuilder.mainSendMessage(Phase4ENTSOGSender.java:218)
	  at com.helger.phase4.sender.AbstractAS4MessageBuilder.sendMessage(AbstractAS4MessageBuilder.java:598)
	  at com.helger.phase4.entsog.MainPhase4PeppolSenderLocalHost8080.main(MainPhase4PeppolSenderLocalHost8080.java:129)
  Caused by: org.apache.wss4j.common.ext.WSSecurityException: No certificates for user "null" were found for signature
	  at org.apache.wss4j.dom.message.WSSecSignature.getSigningCerts(WSSecSignature.java:867)
	  at org.apache.wss4j.dom.message.WSSecSignature.prepare(WSSecSignature.java:171)
	  at org.apache.wss4j.dom.message.WSSecSignature.build(WSSecSignature.java:391)
	  at com.helger.phase4.messaging.crypto.AS4Signer._createSignedMessage(AS4Signer.java:129)
	  at com.helger.phase4.messaging.crypto.AS4Signer.createSignedMessage(AS4Signer.java:189)
	  at com.helger.phase4.client.AS4ClientUserMessage.buildMessage(AS4ClientUserMessage.java:664)
	  at com.helger.phase4.client.AbstractAS4Client.sendMessageWithRetries(AbstractAS4Client.java:458)
	  at com.helger.phase4.sender.AS4BidirectionalClientHelper.sendAS4UserMessageAndReceiveAS4SignalMessage(AS4BidirectionalClientHelper.java:132)
	  at com.helger.phase4.entsog.Phase4ENTSOGSender$AbstractENTSOGUserMessageBuilder.mainSendMessage(Phase4ENTSOGSender.java:196)
	  ... 2 more
  [main] INFO com.helger.phase4.wss.WSSConfigManager - Cleaning up WSSConfig. Security Providers will also be removed.
  
  Process finished with exit code 0
  ```

</p>
</details>

The code that I am trying to use is the following:

```java
final ESuccess eResult;
eResult = Phase4ENTSOGSender.builder()
        // Certificates
        .receiverCertificate(certificate)

        // Setup
        .endpointURL("http://localhost:8080/as4")
        .action("http://docs.oasis-open.org/ebxml-msg/as4/200902/action")
        .service("A06")
        .agreementRef("http://entsog.eu/communication/agreements/" +receiverId + "/" + senderId + "/1")

        // From Partner
        .fromRole(senderRole)
        .fromPartyID(senderId)
        .fromPartyIDType("http://www.entsoe.eu/eic-codes/eic-party-codes-x")

        // To Partner
        .toRole(receiverRole)
        .toPartyID(receiverId)
        .toPartyIDType("http://www.entsoe.eu/eic-codes/eic-party-codes-x")

        // Payload
        .payload(aPayloadElement, params)

        // Send the message
        .sendMessage();
```

Is there an example of the usage of Phase4ENTSOGSender?

If not, can somebody please explain the error more in-depth, because I am currently not sure why the user is "null"

**Comment by aaron-kumar** [comment, marked as answer]
@MakakWasTaken  Please make sure that you have "application.properties" in classpath and following properties configured. 

org.apache.wss4j.crypto.provider=org.apache.wss4j.common.crypto.Merlin
org.apache.wss4j.crypto.merlin.keystore.type=pkcs12
org.apache.wss4j.crypto.merlin.keystore.file=test-ap-2021.p12
org.apache.wss4j.crypto.merlin.keystore.password=peppol
org.apache.wss4j.crypto.merlin.keystore.alias=openpeppol aisbl id von pop000306
org.apache.wss4j.crypto.merlin.keystore.private.password=peppol

org.apache.wss4j.crypto.merlin.load.cacerts=false
#org.apache.wss4j.crypto.merlin.truststore.provider=
org.apache.wss4j.crypto.merlin.truststore.type=jks
org.apache.wss4j.crypto.merlin.truststore.file=truststore/complete-truststore.jks
org.apache.wss4j.crypto.merlin.truststore.password=peppol

**Reply by MakakWasTaken** [reply]
Correct me if I'm wrong, but I thought that you only needed the public certificate to sign and encode the message. 

Why do I then have to have my private key for it to work (.p12)?
And if I only need the public certificate is there a way to define only that? (I thought that was the functionality of receiverCertficate function)

Or do I just need the truststore?

**Reply by phax** [reply]
Most likely, the configuration of the keystore (containing your private key for AS4 (not TLS)) and the truststore (containing the trusted AS4 certificates of your partners)

With your private key, the message is signed, and with the public certificate of the receiver, the message is encrypted.
The receiver will use his private key to decrypt the message (only the receiver can do it) and use your public key to check the signature (check that it is from you).
I am not too familiar with the specific requirements of  ENTSOG regarding the certificate usage.

**Reply by MakakWasTaken** [reply]
Thanks, that clarifies it a lot.

🤓

I'll try to get it working..

**Reply by phax** [reply]
Super. The snippet from Arun contains the property names to use. It is recommended to use PKCS12 keystore and truststore.
Use e.g. the tool "Keystore Explorer" or "Portecle" for managing keystores and truststores with a nice GUI.
Technically keysstore and truststore are the same artefacts - only their "usage" gives them different names. Technically both are (eg.) "PKCS12 Keystores".
hth

**Reply by MakakWasTaken** [reply]
I got one step closer to getting it to work. Now I can actually send the messages between my client and my server.

I have an additional question:

When I am trying to send my file I get the following response error:
`Following errors are present: [PMode.Leg[1].Security.X509SignatureCertificate is missing, PMode.PayloadService is missing]`


I have tried calling the following before sending my message:
```
PMode pmode = ENTSOGPMode.createENTSOGPMode(senderId, receiverId, "http://localhost:8080", IPModeIDProvider.DEFAULT_DYNAMIC, true);
pmode.getLeg1().getSecurity().setX509SignatureCertificate(s_certificate.replaceAll("/\n/g", ""));
```

I am then calling `.pmode(pmode)` on my `Phase4ENTSOGSender.Builder`

This does not seem to do anything, is there anything else that I need to do so that my PMode parameters get added?

**Reply by phax** [reply]
You should not set the PMode yourself - that should be hidden in the background.
Most likely you just need to call the `receiverCertificate` on the builder, passing in your certificate with the private key.

**Reply by MakakWasTaken** [reply]
I have set the receiver certificate on the Builder. But that did not Seem to do anything, As the argument I am passing the CertficateHelper convert function, with the public certificate.

I am currently not at my computer, should I use the private certificate of the receiver instead instead? Does this not defeat the purpose of private and public certificates?

**Reply by phax** [reply]
You are right - I am confused myself. The `receiverCertificate` is the public one that will be used to encrypt the message.
As I am not using ENTSOG myself, I need to contact the author to verify how this is intended....

**Reply by phax** [reply]
Okay, I relaxed the checks in the ENTSOG pmode profile validator, so that your issues should be resolved.
Please check the latest 1.3.9-SNAPSHOT version, and if it works, I will create an updated release.

**Reply by MakakWasTaken** [reply]
That did it, now I am able to send the message back and forth and the payload is received.

**Reply by phax** [reply]
Okay cool - I will release an 1.3.9 version asap

**Reply by phax** [reply]
Release phase4 1.3.9 is released and soon on Maven Central
